cpp_tests/test_Test.cpp

```diff
@@ -0,0 +1,417 @@
+	Test_arraysC data;
+	remote->getSample_arrays(&data);
+	arraysEqual<Test_arraysC>(expected_data, &data);
+
+	// at this point there should be nothing on the queue
+	REQUIRE(remote->getNextSample_arrays(&data) == SAL__NO_UPDATES);
+	REQUIRE(remote->getSample_arrays(&data) == SAL__NO_UPDATES);
+    }
+
+    /**
+     * Test that get newest after get oldest gets the newest value.
+     * <p>
+     * This tests DM-18491.
+     */
+    SECTION("EvtGetNewestAfterGetOldest") {
+	remote->salEventSub((char *)"Test_logevent_arrays");
+	controller->salEventPub((char *)"Test_logevent_arrays");
+
+	getNewestAfterGetOldest<Test_logevent_arraysC>(
+		std::bind(&SAL_Test::logEvent_arrays, controller, _1, 1),
+		std::bind(&SAL_Test::getNextSample_logevent_arrays, remote, _1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1));
+	getNewestAfterGetOldest<Test_logevent_arraysC>(
+		std::bind(&SAL_Test::logEvent_arrays, controller, _1, 1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1),
+		std::bind(&SAL_Test::getSample_logevent_arrays, remote, _1), true);
+    }
+
+    /**
+     * Test that get newest after get oldest gets the newest value.
+     * <p>
+     * This tests DM-18491.
+     */
+    SECTION("TelGetNewestAfterGetOldest") {
+	remote->salTelemetrySub((char *)"Test_arrays");
+	controller->salTelemetryPub((char *)"Test_arrays");
+
+	getNewestAfterGetOldest<Test_arraysC>(std::bind(&SAL_Test::putSample_arrays, controller, _1),
+					      std::bind(&SAL_Test::getNextSample_arrays, remote, _1),
+					      std::bind(&SAL_Test::getSample_arrays, remote, _1));
+    }
+
+    SECTION("TelGetNewestAfterGetOldest") {
+	remote->salTelemetrySub((char *)"Test_arrays");
+	controller->salTelemetryPub((char *)"Test_arrays");
+
+	getNewestAfterGetOldest<Test_arraysC>(std::bind(&SAL_Test::putSample_arrays, controller, _1),
+					      std::bind(&SAL_Test::getNextSample_arrays, remote, _1),
+					      std::bind(&SAL_Test::getSample_arrays, remote, _1));
+    }
+
+    /**
+     * Test that a late joiner can can read the most recent event using getNextSample.
+     * <p>
+     * Only one value is retrievable.
+     */
+    SECTION("EvtLateJoinerOldest") {
+	controller->salEventPub((char *)"Test_logevent_arrays");
+
+	Test_logevent_arraysC dataList[maxLoops];
+	for (int i = 0; i < maxLoops; i++) {
+	    Test_logevent_arraysC data;
+	    fillArraysWithRandomValues<Test_logevent_arraysC>(&data);
+	    dataList[i] = data;
+	    REQUIRE(controller->logEvent_arrays(&data, 1) == SAL__OK);
+	}
+
+	Test_logevent_arraysC data;
+	remote->salEventSub((char *)"Test_logevent_arrays");
+
+	REQUIRE(remote->getNextSample_logevent_arrays(&data) == SAL__OK);
+	arraysEqual<Test_logevent_arraysC>(&data, &(dataList[0]));
+    }
+
+    /**
+     * Test that a late joiner cannot see historical telemetry using getNextSample.
+     * <p>
+     * Telemetry is volatile so there should be no late joiner data.
+     */
+    SECTION("TelLateJoinerOldest") {
```

Comment:
I gently suggest Oldest -> getNextSample because it's clearer exactly what is being tested (and because events have 2 ways to get the oldest data). I've changed the Java and Python.